central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -202,45 +293,144 @@ func (ds *datastoreImpl) GetBatch(ctx context.Context, ids []string) ([]*storage
 	return collections, nil
 }
 
-func (ds *datastoreImpl) AddCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+func (ds *datastoreImpl) addCollectionWorkflow(ctx context.Context, collection *storage.ResourceCollection, dryrun bool) error {
+
+	// sanity checks
+	if err := verifyCollectionObjectNotEmpty(collection); err != nil {
+		return err
+	}
+	if collection.GetId() != "" {
+		return fmt.Errorf("added collections must not have an `id` preset (%s)", collection.GetId())
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
 
-	// add to graph first to detect any cycles
-	err := ds.addCollectionToGraph(collection)
+	// verify that the name is not already in use
+	if collection.GetName() == "" || ds.names.Contains(collection.GetName()) {
+		return fmt.Errorf("collections must have non-empty, unique `name` values (%s)", collection.GetName())
+	}
+
+	// add to graph to detect any cycles, this also sets the `id` field
+	graph, err := ds.addCollectionToGraph(collection)
 	if err != nil {
 		return err
 	}
 
+	// if this is a dryrun, we don't want to add to storage or make changes to objects
+	if dryrun {
+		collection.Id = ""
+		return nil
+	}
+
+	// add to storage
 	err = ds.storage.Upsert(ctx, collection)
 	if err != nil {
-		// if we fail to upsert, update the graph
-		deleteErr := ds.deleteCollectionFromGraph(collection.GetId())
-		if deleteErr != nil {
-			log.Errorf("Failed to remove collection from internal state object (%v)", deleteErr)
-		}
 		return err
 	}
-	return err
+
+	// we've succeeded, now set all the values
+	ds.names.Add(collection.GetName())
+	ds.graph = graph
+	return nil
+}
+
+func (ds *datastoreImpl) AddCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+	return ds.addCollectionWorkflow(ctx, collection, false)
```

Comment:
check sac

Reply:
This operation will not succeed unless it passes a call to postgres' `Upsert`, which will do the sac check for us

Reply:
The pre-req steps, which are blocking in nature, are executed un-necessarily before the postgres check happens. I strongly recommend the workflows be such that they avoid blocking when possible.